src/main/scala/calc/TypeChecker.scala

```diff
@@ -0,0 +1,52 @@
+package calc
+
+import org.scalajs.core.ir
+import ir.{Trees => irt, Types => irtpe}
+import ir.Definitions._
+import collection.immutable.Map
+
+sealed abstract class TreeType
+
+case object DoubleType extends TreeType
+
+case class FunType(numOfArgs: Int) extends TreeType
+
+/**
+  * Created by ivan on 16/3/23.
+  * A type checker and infer about types
+  */
+object TypeChecker {
+  def typeInfer(tree: Tree, typeEnv: Map[String, TreeType] = Map()): TreeType ={
+    tree match {
+      case Literal(value) => DoubleType
+      case Ident(name) => typeEnv(name)
```

Comment:
This will throw a not-very-nice exception for the developer if the identifier is not in the scope.